whelktool/scripts/cleanups/2022/02/lxl-3658-link-materials.groovy

```diff
@@ -0,0 +1,67 @@
+PrintWriter linked = getReportWriter('linked.txt')
+PrintWriter changedType = getReportWriter('changed-type.txt')
+
+MATERIAL_BASE_URI = 'https://id.kb.se/material/'
+
+LABEL_MAPPINGS =
+        [
+                'Paper'      : ~/pap[pi]?er/,
+                'Parchment'  : ~/pergament/,
+                'Plastic'    : ~/plast(ic)?/,
+                'Glass'      : ~/glas/,
+                'Metal'      : ~/metall?/,
+                'Canvas'     : ~/målarduk/,
+                'Textile'    : ~/textile?/,
+                'Vinyl'      : ~/vinyl/,
+                'Plaster'    : ~/gips/,
+                'Watercolour': ~/akvarell|vattenfärg/,
+                'Gouache'    : ~/gouas?che?/,
+                'Charcoal'   : ~/kol/,
+                'Ink'        : ~/bläck/,
+                'OilPaint'   : ~/oljefärg|olja/,
+                'Synthetic'  : ~/syntetiskt material/,
+                'Lacquer'    : ~/fernissa/,
+                'Chalk'      : ~/krita/,
+                'Cardboard'  : ~/papp|kartong/,
+                'Pastel'     : ~/torrpastell/
+        ]
+
+['baseMaterial', 'appliedMaterial'].each { property ->
+    selectByIds(queryIds([('exists-' + property): ['true']]).collect()) { data ->
+        def id = data.doc.shortId
+        def thing = data.graph[1]
+        def modified = false
+        def add = [] as Set
+
+        thing[property].removeAll { m ->
+            if (m.'@type') {
+                def foundLinks = findLinks(m)
+                if (foundLinks) {
+                    linked.println("$id\t${m}\t${foundLinks}")
+                    add += foundLinks
+                    return modified = true
+                }
+                changedType.println("$id\t${m}")
+                m['@type'] = 'Material'
+                modified = true
+            }
+            return false
+        }
+
+        thing[property] += add.collect { ['@id': MATERIAL_BASE_URI + it] }
+
+        if (modified)
+            data.scheduleSave()
+    }
+}
+
+List findLinks(Map material) {
+    def label = material.label in List ? material.label.join(' & ') : material.label
```

Comment:
Is `in List` supposed to be `instanceof List` mayhaps?

Reply:
Just sugar, albeit not with the most obvious meaning.

Reply:
TIL! (I suppose it's set-theoretically obvious. :slightly_smiling_face: )